src/esp/scene/SceneNode.cpp

```diff
@@ -12,26 +14,171 @@ namespace scene {
   node->setId(this->getId());
+  node->setSceneNodeTags(childNodeTags);
   return *node;
 }
 
+SceneNode& SceneNode::setParent(SceneNode* newParent) {
+  // Perform same internal checks as magnum to ensure newParent is a valid new
+  // parent before updating any SensorSuites
+  // Parent can not be leaf node
+  CORRADE_ASSERT(!(newParent->getSceneNodeTags() & SceneNodeTag::Leaf),
+                 "SceneNode::setParent(): New parent node can not be leaf node",
+                 *this);
+
+  // Skip if this is scene or RootNode (which cannot have parent), or if
+  // newParent is already parent
+  if (isScene() || SceneGraph::isRootNode(*this) ||
+      this->parent() == newParent) {
+    return *this;
+  }
+
+  // Skip if this is an ancestor of newParent
+  auto p = newParent->parent();
+  while (p) {
+    if (p == this)
+      return *this;
+    p = p->parent();
+  }
+
+  // Update old ancestors' SubtreeSensorSuites
+  removeSubtreeSensorsFromAncestors();
+  // Remove sensors from old parent node's nodeSensorSuite
+  removeSensorFromParentNodeSensorSuite();
+
+  MagnumObject::setParent(newParent);
+
+  // Update new ancestors'SubtreeSensorSuites
+  addSubtreeSensorsToAncestors();
+  // Add sensors to newParent's nodeSensorSuite
+  addSensorToParentNodeSensorSuite();
+  return *this;
+}
+
+void SceneNode::addSensorToParentNodeSensorSuite() {
+  // Only add a sensor if this is a leaf node and sensor exists
+  if (!(getSceneNodeTags() & SceneNodeTag::Leaf) ||
+      nodeSensorSuite_->getSensors().empty()) {
+    return;
+  }
+  // There only exists 0 or 1 sensors in a leaf nodeSensorSuite
+  // Get the first one, and if it is valid, add it to parent's nodeSensorSuite
+  CORRADE_INTERNAL_ASSERT(nodeSensorSuite_->getSensors().size() == 1);
+  std::map<std::string, std::reference_wrapper<sensor::Sensor>>::iterator it =
+      nodeSensorSuite_->getSensors().begin();
+  if (it != nodeSensorSuite_->getSensors().end()) {
+    CORRADE_ASSERT(static_cast<scene::SceneNode*>(this->parent()) != nullptr,
+                   "SceneNode::addSensorToParentNodeSensorSuite(): Parent null "
+                   "is nullptr", );
+    static_cast<scene::SceneNode*>(this->parent())
+        ->getNodeSensorSuite()
+        .add(it->second);
+  }
+}
+
+void SceneNode::removeSensorFromParentNodeSensorSuite() {
+  // Only remove a sensor if this is a leaf node and sensor exists
+  if (!(getSceneNodeTags() & SceneNodeTag::Leaf) ||
+      nodeSensorSuite_->getSensors().empty()) {
+    return;
+  }
+  // There only exists 0 or 1 sensors in a leaf nodeSensorSuite
+  // Get the first one, and if it is valid, remove it from parent's
+  // nodeSensorSuite
+  CORRADE_INTERNAL_ASSERT(nodeSensorSuite_->getSensors().size() == 1);
+  std::map<std::string, std::reference_wrapper<sensor::Sensor>>::iterator it =
+      nodeSensorSuite_->getSensors().begin();
+  if (it != nodeSensorSuite_->getSensors().end()) {
+    CORRADE_ASSERT(static_cast<scene::SceneNode*>(this->parent()) != nullptr,
```

Comment:
Here, the same thing. `static_cast` is the problem.
And you just simply copy and paste whatever you have in the `addSensorToParentNodeSensorSuite`.
**But clearly the logic is different in an "add" and a "remove" functions.**
Here, if you find you cannot dynamic_cast to its parent, one case is that the parent (SceneNode) is already deconstructed. So the conversion fails., which means in this case, you do not need to do anything. (The logic you have here is to let the program exit due to ASSERTION failure. It is totally wrong.)

Reply:
Ok, thanks for the clarification!